src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+        self.tmp_file.write("[remote]\n")
+        for k, v in config.items():
+            self.tmp_file.write(f"{k} = {v}\n")
+
+        self.tmp_file.flush()
+
+        return RcloneConfigTuple(self.tmp_file.name, remote_path)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.tmp_file:
+            self.tmp_file.close()
+
+
+async def rclone(job, cloud_sync):
+    # Use a temporary file to store rclone file
+    with RcloneConfig(cloud_sync) as config:
+        args = [
+            '/usr/local/bin/rclone',
+            '--config', config.config_path,
+            '-v',
+            '--stats', '1s',
+            cloud_sync['transfer_mode'].lower(),
+        ]
+
+        if cloud_sync['direction'] == 'PUSH':
+            args.extend([cloud_sync['path'], config.remote_path])
+        else:
+            args.extend([config.remote_path, cloud_sync['path']])
+
+        proc = await Popen(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+        check_cloud_sync = asyncio.ensure_future(rclone_check_progress(job, proc))
+        await proc.wait()
+        if proc.returncode != 0:
+            await asyncio.wait_for(check_cloud_sync, None)
+            raise ValueError("rclone failed")
+        return True
+
+
+async def rclone_check_progress(job, proc):
+    RE_TRANSF = re.compile(r'Transferred:\s*?(.+)$', re.S)
+    while True:
+        read = (await proc.stdout.readline()).decode()
+        job.logs_fd.write(read.encode("utf-8", "ignore"))
+        if read == '':
+            break
+        reg = RE_TRANSF.search(read)
+        if reg:
+            transferred = reg.group(1).strip()
+            if not transferred.isdigit():
+                job.set_progress(None, transferred)
+
+
+def rclone_encrypt_password(password):
+    key = bytes([0x9c, 0x93, 0x5b, 0x48, 0x73, 0x0a, 0x55, 0x4d,
+                 0x6b, 0xfd, 0x7c, 0x63, 0xc8, 0x86, 0xa9, 0x2b,
+                 0xd3, 0x90, 0x19, 0x8e, 0xb8, 0x12, 0x8a, 0xfb,
+                 0xf4, 0xde, 0x16, 0x2b, 0x8b, 0x95, 0xf6, 0x38])
+
+    iv = Random.new().read(AES.block_size)
+    counter = Counter.new(128, initial_value=int(codecs.encode(iv, "hex"), 16))
+    cipher = AES.new(key, AES.MODE_CTR, counter=counter)
+    encrypted = iv + cipher.encrypt(password.encode("utf-8"))
+    return base64.urlsafe_b64encode(encrypted).decode("ascii").rstrip("=")
+
+
+def validate_attributes(schema, data, additional_attrs=False):
+    verrors = ValidationErrors()
+
+    schema = Dict("attributes", *schema, additional_attrs=additional_attrs)
+
+    try:
+        data["attributes"] = schema.clean(data["attributes"])
+    except Error as e:
+        verrors.add(e.attribute, e.errmsg, e.errno)
+
+    try:
+        schema.validate(data["attributes"])
+    except ValidationErrors as e:
+        verrors.extend(e)
+
+    return verrors
+
+
+class CredentialsService(CRUDService):
+
+    class Config:
+        namespace = 'cloudsync.credentials'
```

Comment:
Can we use the datastore attribute here? Then we could get rid of query method entirely, using the default implementation from CRUDService.

Reply:
Fixed